src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp

```diff
@@ -233,19 +235,36 @@ NOINLINE frame os::current_frame() {
 
 bool PosixSignals::pd_hotspot_signal_handler(int sig, siginfo_t* info,
                                              ucontext_t* uc, JavaThread* thread) {
-  // Enable WXWrite: this function is called by the signal handler at arbitrary
-  // point of execution.
-  ThreadWXEnable wx(WXWrite, thread);
-
   // decide if this trap can be handled by a stub
   address stub = nullptr;
-
-  address pc          = nullptr;
+  address pc   = nullptr;
 
   //%note os_trap_1
   if (info != nullptr && uc != nullptr && thread != nullptr) {
     pc = (address) os::Posix::ucontext_get_pc(uc);
 
+#ifdef MACOS_W_XOR_X
+    // If we got a SIGBUS because we tried to write into the code
+    // cache, try enabling WXWrite mode.
+    if (sig == SIGBUS && pc != info->si_addr && CodeCache::contains(info->si_addr)) {
+      WXMode *entry_mode = thread->_cur_wx_mode;
+      if (*entry_mode == WXArmedForWrite) {
+        if (TraceWXHealing) {
+          static const char *mode_names[3] = {"WXWrite", "WXExec", "WXArmedForWrite"};
+          tty->print_cr("Healing WXMode %s at %p to WXWrite",
+                        mode_names[*entry_mode], entry_mode);
+        }
+        *(thread->_cur_wx_mode) = WXWrite;
+      }
+      return thread->wx_enable_write();
+    }
+
+    // Enable WXWrite for the duration of this handler: this function
+    // is called by the signal handler at arbitrary point of
+    // execution.
+    ThreadWXEnable wx(WXWrite, thread);
```

Comment:
I'm not sure this ThreadWXEnable needs to use WXWrite.  I had changed it to WXExec in my experiments.